rules/firestore.json

```diff
@@ -35,50 +37,204 @@ service cloud.firestore {
+            allow create: if isOwner(userId) && request.resource.data.verified == false
+            // Only the user can change its fields
+            // Moreover, the username cannot be changed once set
+            // And the verified field can only be changed to true if the user has a username and has its email verified
+            allow update: if isOwner(userId) && doesNotModifyUsernameIfSet() && onlyModifiesVerifiedFieldIfUserIsVerified();
+            // Anybody can read the fields of a user.
+            // This is because before user creation, we need to check if the username is free.
             allow read: if true;
         }
-        match /parties/{id} {
-            allow write, read: if userVerified();
-        }
-        match /joiners/{id} {
-            allow write, read: if userVerified();
+        function isOwner(userId) {
+            return request != null &&
+                request.auth != null &&
+                request.auth.uid == userId;
         }
-        function userVerified() {
-            return isAuthentified() &&
+        function doesNotModifyUsernameIfSet() {
+            let usernameAlreadySet = hasField(resource.data, 'username');
+            let usernameBeingSet = hasField(request.resource.data, 'username');
+            // It sometimes happens that firebase sends the username as part as updates, even though it is not actually updated. Hence, we accept usernames as part of updates, only if it doesn't change.
+            let doesNotChange = usernameAlreadySet && usernameBeingSet && resource.data['username'] == request.resource.data['username'];
+            return usernameAlreadySet == false || usernameBeingSet == false || doesNotChange;
+        }
+        function onlyModifiesVerifiedFieldIfUserIsVerified() {
+            let verifiedKeptSameValue = (resource.data.verified == request.resource.data.verified);
+            let verifiedSetToTrueAfterVerification =  (resource.data.verified == false && userIsVerified() && resource.data.username != null);
+            return verifiedKeptSameValue || verifiedSetToTrueAfterVerification
+        }
+        match /parts/{id} {
+            // Any verified user can create a part with the proper fields set, and has to be the creator
+            allow create: if userIsVerified() &&
+                             // turn can be -1 in case of a new part, 0 in case of a rematch
+                             (request.resource.data.turn == -1 || request.resource.data.turn == 0) &&
+                             hasField(request.resource.data, 'typeGame') &&
+                             request.resource.data.listMoves == [] &&
+                             // creator must be one of the players (could be playerOne in case of rematch)
+                             (request.auth.uid == request.resource.data.playerZero.id ||
+                              hasField(request.resource.data, 'playerOne') && request.auth.uid == request.resource.data.playerOne.id);
```

Comment:
p-e pas le mettre dans une fonction, à toi de voir, mais par contre "a or b and c" j'dois être honnête j'suis pas rachuré lo

Reply:
mieux avec une parenthèse ? (a or (b and c))
Même si PEMDAS ! Ou plutôt PAO dans le monde logique :-'

Reply:
PAO pour moi c'est le cri qu'on pousse quand on donne un coup de pied au TaeKwonDo, va me falloir de l'aide 🗡️

Reply:
Si t'as un exemple de langage qui a pas l'ordre usuel pour les opérations logiques je veux bien voir pour cracher dessus :p Mais d'un point de vue lecture c'est mieux avec les parenthèses, donc ajoutées !